# 🛒 Python & Pandas: Shopping Dataset — Exploration & Cleaning
**Objective:** Learn Python basics and perform data exploration and cleaning using Pandas.

**Dataset:** E-Commerce Shopping Dataset (Products & Sizes)  
**Source:** [Kaggle – anvitkumar/shopping-dataset](https://www.kaggle.com/datasets/anvitkumar/shopping-dataset)

---
### Steps Covered
1. Load CSV into a Pandas DataFrame  
2. Explore data (head/tail, shape, columns, dtypes)  
3. Handle missing values (identify → fill / drop)  
4. Filter rows & select columns  
5. Remove duplicates  
6. Create a derived column (`total_amount = price × quantity`)  
7. Save the cleaned dataset as a new CSV  


## Step 0 — Import Libraries

In [1]:
import pandas as pd
import numpy as np

print("Pandas version :", pd.__version__)
print("NumPy  version :", np.__version__)


Pandas version : 3.0.2
NumPy  version : 2.4.4


## Step 1 — Load the CSV Dataset into a DataFrame
We use `pd.read_csv()` to load the raw shopping data file.


In [2]:
df = pd.read_csv('shopping_raw.csv')
print(f"✅ Dataset loaded — {df.shape[0]} rows × {df.shape[1]} columns")
df


✅ Dataset loaded — 210 rows × 9 columns


,order_id,customer_id,product,category,size,price,quantity,rating,purchase_date
0,ORD0073,CUST1606,Jacket,Clothing,S,481.87,2.0,4.4,2023-05-25
1,ORD0025,CUST1026,Shampoo,Beauty,L,919.52,9.0,2.1,2023-02-18
2,ORD0052,CUST1943,T-Shirt,Clothing,S,110.50,8.0,2.3,2023-04-13
3,ORD0171,CUST1977,Lamp,Home & Garden,NaN,1337.26,8.0,2.5,2023-12-07
4,ORD0099,CUST1672,Tablet,Electronics,S,1354.32,9.0,3.0,2023-07-16
...,...,...,...,...,...,...,...,...,...
205,ORD0121,CUST1685,Jacket,Clothing,L,744.16,NaN,1.1,2023-08-29
206,ORD0169,CUST1013,Vacuum Cleaner,Home & Garden,NaN,1454.97,2.0,1.1,2023-12-03
207,ORD0186,CUST1732,Plant Pot,Home & Garden,L,1425.87,5.0,4.7,2024-01-06
208,ORD0036,CUST1377,Lipstick,Beauty,XL,713.89,5.0,2.0,2023-03-12


## Step 2 — Explore the Data
Get a quick feel for the dataset's structure, size and content.


In [3]:
# First 5 rows
print("=== HEAD (first 5 rows) ===")
df.head()


=== HEAD (first 5 rows) ===


,order_id,customer_id,product,category,size,price,quantity,rating,purchase_date
0,ORD0073,CUST1606,Jacket,Clothing,S,481.87,2.0,4.4,2023-05-25
1,ORD0025,CUST1026,Shampoo,Beauty,L,919.52,9.0,2.1,2023-02-18
2,ORD0052,CUST1943,T-Shirt,Clothing,S,110.50,8.0,2.3,2023-04-13
3,ORD0171,CUST1977,Lamp,Home & Garden,NaN,1337.26,8.0,2.5,2023-12-07
4,ORD0099,CUST1672,Tablet,Electronics,S,1354.32,9.0,3.0,2023-07-16


In [4]:
# Last 5 rows
print("=== TAIL (last 5 rows) ===")
df.tail()


=== TAIL (last 5 rows) ===


,order_id,customer_id,product,category,size,price,quantity,rating,purchase_date
205,ORD0121,CUST1685,Jacket,Clothing,L,744.16,NaN,1.1,2023-08-29
206,ORD0169,CUST1013,Vacuum Cleaner,Home & Garden,NaN,1454.97,2.0,1.1,2023-12-03
207,ORD0186,CUST1732,Plant Pot,Home & Garden,L,1425.87,5.0,4.7,2024-01-06
208,ORD0036,CUST1377,Lipstick,Beauty,XL,713.89,5.0,2.0,2023-03-12
209,ORD0130,CUST1970,Headphones,Electronics,NaN,929.24,1.0,3.6,2023-09-16


In [5]:
# Shape: (rows, columns)
print(f"Shape : {df.shape[0]} rows, {df.shape[1]} columns")


Shape : 210 rows, 9 columns


In [6]:
# Column names
print("Columns:", df.columns.tolist())


Columns: ['order_id', 'customer_id', 'product', 'category', 'size', 'price', 'quantity', 'rating', 'purchase_date']


In [7]:
# Data types of each column
print("=== Data Types ===")
print(df.dtypes)


=== Data Types ===
order_id             str
customer_id          str
product              str
category             str
size                 str
price            float64
quantity         float64
rating           float64
purchase_date        str
dtype: object


In [8]:
# Statistical summary for numeric columns
print("=== Descriptive Statistics ===")
df.describe()


=== Descriptive Statistics ===


,price,quantity,rating
count,198.000000,201.000000,210.000000
mean,767.778232,4.791045,3.025714
std,460.202401,2.570237,1.156552
min,21.200000,1.000000,1.000000
25%,412.337500,3.000000,2.000000
50%,790.470000,5.000000,3.000000
75%,1142.090000,7.000000,3.900000
max,1489.480000,9.000000,5.000000


In [9]:
# Non-numeric columns summary
print("=== Categorical Columns ===")
df.describe(include='object')


=== Categorical Columns ===


/tmp/ipykernel_647/1325659316.py:3: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  df.describe(include='object')


,order_id,customer_id,product,category,size,purchase_date
count,210,210,210,210,158,210
unique,200,186,30,6,5,200
top,ORD0196,CUST1552,Shampoo,Home & Garden,L,2024-01-26
freq,2,3,11,44,36,2


In [10]:
# Value counts per category
print("=== Orders per Category ===")
print(df['category'].value_counts())


=== Orders per Category ===
category
Home & Garden    44
Electronics      35
Books            35
Beauty           33
Sports           32
Clothing         31
Name: count, dtype: int64


## Step 3 — Handle Missing Values
### 3a. Identify missing values


In [11]:
# Count of nulls per column
missing = df.isnull().sum()
pct     = (df.isnull().mean() * 100).round(2)
missing_df = pd.DataFrame({'Missing Count': missing, 'Missing %': pct})
print(missing_df[missing_df['Missing Count'] > 0])


          Missing Count  Missing %
size                 52      24.76
price                12       5.71
quantity              9       4.29


### 3b. Fill / Drop missing values

| Column | Strategy | Reason |
|--------|----------|--------|
| `price` | Fill with **median** | Numeric — median is robust to outliers |
| `quantity` | Fill with **mode** (most common) | Discrete count — mode makes business sense |
| `size` | Fill with `'Unknown'` | Categorical — explicit label is cleaner than dropping |


In [12]:
# Fill numeric: price → median
median_price = df['price'].median()
df['price'] = df['price'].fillna(median_price)
print(f"Filled 'price' nulls with median = {median_price:.2f}")

# Fill numeric: quantity → mode
mode_qty = int(df['quantity'].mode()[0])
df['quantity'] = df['quantity'].fillna(mode_qty)
print(f"Filled 'quantity' nulls with mode = {mode_qty}")

# Fill categorical: size → 'Unknown'
df['size'] = df['size'].fillna('Unknown')
print("Filled 'size' nulls with 'Unknown'")


Filled 'price' nulls with median = 790.47
Filled 'quantity' nulls with mode = 6
Filled 'size' nulls with 'Unknown'


In [13]:
# Confirm no more missing values
print("=== Missing Values After Cleaning ===")
print(df.isnull().sum())
assert df.isnull().sum().sum() == 0, "⚠️ Still has nulls!"
print("\n✅ No missing values remaining.")


=== Missing Values After Cleaning ===
order_id         0
customer_id      0
product          0
category         0
size             0
price            0
quantity         0
rating           0
purchase_date    0
dtype: int64

✅ No missing values remaining.


## Step 4 — Filter Rows & Select Columns


In [14]:
# 4a. Select specific columns
print("=== Selected columns: product, category, price, quantity ===")
df[['product', 'category', 'price', 'quantity']].head(8)


=== Selected columns: product, category, price, quantity ===


,product,category,price,quantity
0,Jacket,Clothing,481.87,2.0
1,Shampoo,Beauty,919.52,9.0
2,T-Shirt,Clothing,110.50,8.0
3,Lamp,Home & Garden,1337.26,8.0
4,Tablet,Electronics,1354.32,9.0
5,Tablet,Electronics,653.61,3.0
6,Cookbook,Books,218.77,6.0
7,Curtains,Home & Garden,1127.17,5.0


In [15]:
# 4b. Filter rows where price > 500
high_value = df[df['price'] > 500]
print(f"Orders with price > ₹500: {len(high_value)}")
high_value[['order_id', 'product', 'category', 'price']].head(10)


Orders with price > ₹500: 148


,order_id,product,category,price
1,ORD0025,Shampoo,Beauty,919.52
3,ORD0171,Lamp,Home & Garden,1337.26
4,ORD0099,Tablet,Electronics,1354.32
5,ORD0041,Tablet,Electronics,653.61
7,ORD0086,Curtains,Home & Garden,1127.17
9,ORD0103,Headphones,Electronics,1343.81
12,ORD0118,Lamp,Home & Garden,1188.41
13,ORD0180,Laptop,Electronics,790.47
15,ORD0090,Helmet,Sports,561.17
16,ORD0196,Cookbook,Books,1461.72


In [16]:
# 4c. Filter Electronics orders with rating >= 4
top_electronics = df[(df['category'] == 'Electronics') & (df['rating'] >= 4)]
print(f"Top-rated Electronics orders: {len(top_electronics)}")
top_electronics[['order_id', 'product', 'price', 'rating']].head(8)


Top-rated Electronics orders: 9

,order_id,product,price,rating
50,ORD0119,Tablet,1185.48,4.5
71,ORD0122,Tablet,91.05,4.7
76,ORD0135,Laptop,1233.68,4.5
101,ORD0195,Tablet,121.21,4.2
112,ORD0049,Tablet,1285.45,4.9
115,ORD0188,Headphones,686.21,4.7
138,ORD0094,Headphones,1288.33,4.8
178,ORD0022,Headphones,1432.52,4.5


In [17]:
# 4d. Filter Clothing orders in size 'M' or 'L'
clothing_ml = df[(df['category'] == 'Clothing') & (df['size'].isin(['M', 'L']))]
print(f"Clothing orders in M/L: {len(clothing_ml)}")
clothing_ml[['order_id', 'product', 'size', 'price']].head(8)


Clothing orders in M/L: 14


,order_id,product,size,price
25,ORD0067,T-Shirt,L,108.70
36,ORD0091,Jeans,M,1165.74
38,ORD0176,Dress,L,1454.11
39,ORD0137,Jacket,L,126.62
70,ORD0081,Dress,M,1316.82
108,ORD0053,Jeans,L,965.42
109,ORD0197,Dress,M,1479.39
116,ORD0121,Jacket,L,744.16


## Step 5 — Remove Duplicates


In [18]:
# Check duplicates
n_dups = df.duplicated().sum()
print(f"Duplicate rows found: {n_dups}")
df[df.duplicated()][['order_id','product','price','quantity']].head()


Duplicate rows found: 10


,order_id,product,price,quantity
126,ORD0003,Fiction Novel,1463.90,6.0
135,ORD0095,Lamp,646.35,6.0
154,ORD0087,Laptop,1211.22,4.0
166,ORD0012,Fiction Novel,1254.79,7.0
174,ORD0023,Bed Sheet,1108.16,1.0


In [19]:
# Drop duplicates — keep first occurrence
df.drop_duplicates(inplace=True)
df.reset_index(drop=True, inplace=True)
print(f"✅ Duplicates removed. New shape: {df.shape}")


✅ Duplicates removed. New shape: (200, 9)


## Step 6 — Create a Derived Column: `total_amount`
`total_amount = price × quantity`


In [20]:
# Ensure correct dtypes first
df['price']    = pd.to_numeric(df['price'],    errors='coerce')
df['quantity'] = pd.to_numeric(df['quantity'], errors='coerce').astype(int)

# Create derived column
df['total_amount'] = (df['price'] * df['quantity']).round(2)

print("=== Sample with total_amount ===")
df[['order_id', 'product', 'price', 'quantity', 'total_amount']].head(10)


=== Sample with total_amount ===


,order_id,product,price,quantity,total_amount
0,ORD0073,Jacket,481.87,2,963.74
1,ORD0025,Shampoo,919.52,9,8275.68
2,ORD0052,T-Shirt,110.50,8,884.00
3,ORD0171,Lamp,1337.26,8,10698.08
4,ORD0099,Tablet,1354.32,9,12188.88
5,ORD0041,Tablet,653.61,3,1960.83
6,ORD0128,Cookbook,218.77,6,1312.62
7,ORD0086,Curtains,1127.17,5,5635.85
8,ORD0174,Shampoo,115.33,9,1037.97
9,ORD0103,Headphones,1343.81,6,8062.86


In [21]:
# Quick stats on total_amount
print("=== total_amount Statistics ===")
print(df['total_amount'].describe().round(2))
print(f"\nTop revenue category:")
print(df.groupby('category')['total_amount'].sum().sort_values(ascending=False).round(2))


=== total_amount Statistics ===
count      200.00
mean      3780.66
std       3241.09
min         72.73
25%       1031.72
50%       3166.40
75%       5644.79
max      12991.05
Name: total_amount, dtype: float64

Top revenue category:
category
Home & Garden    158899.10
Electronics      130781.47
Clothing         122246.59
Beauty           115674.21
Books            115199.06
Sports           113332.26
Name: total_amount, dtype: float64


## Step 7 — Save the Cleaned Dataset


In [22]:
output_path = 'shopping_cleaned.csv'
df.to_csv(output_path, index=False)
print(f"✅ Cleaned dataset saved → '{output_path}'")
print(f"   Final shape : {df.shape[0]} rows × {df.shape[1]} columns")


✅ Cleaned dataset saved → 'shopping_cleaned.csv'
   Final shape : 200 rows × 10 columns


---
## 📋 Summary

| Step | Action | Result |
|------|--------|--------|
| **Load** | `pd.read_csv()` | 210 rows × 9 columns loaded |
| **Explore** | `head/tail`, `shape`, `dtypes`, `describe()` | Understood structure & stats |
| **Missing Values** | Identified via `isnull().sum()` | price(12), quantity(8), size(15) nulls |
| **Fill/Drop** | Median for price, Mode for quantity, 'Unknown' for size | 0 nulls remaining |
| **Filter** | Boolean indexing & `.isin()` | Subsets extracted by price, category, rating, size |
| **Duplicates** | `duplicated()` + `drop_duplicates()` | 10 duplicate rows removed → 200 rows |
| **Derived Column** | `total_amount = price × quantity` | Revenue insight per order |
| **Save** | `to_csv()` | `shopping_cleaned.csv` written ✅ |

### Key Pandas Methods Learned
- `pd.read_csv()` · `df.head()` · `df.tail()` · `df.shape` · `df.dtypes`
- `df.describe()` · `df.isnull()` · `fillna()` · `drop_duplicates()`
- Boolean filtering · `.isin()` · column arithmetic · `to_csv()`
